climada/hazard/base.py

```diff
@@ -369,6 +377,415 @@ def set_vector(self, *args, **kwargs):
+        >>>         latitude=[0, 1],
+        >>>         longitude=[0, 1, 2],
+        >>>     ),
+        >>> )
+        >>> hazard = Hazard.from_raster_xarray(
+        >>>     dset,
+        >>>     "",
+        >>>     "",
+        >>>     data_vars=dict(
+        >>>         # Load frequency from 'freq' array
+        >>>         frequency="freq",
+        >>>         # Ignore 'event_id' array and use default instead
+        >>>         event_id="",
+        >>>         # 'fraction' array is loaded because it has the default name
+        >>>     ),
+        >>> )
+        >>> np.array_equal(hazard.frequency, [0.4]) and np.array_equal(
+        >>>     hazard.event_id, [1]
+        >>> )
+        True
+
+        """
+        # If the data is a string, open the respective file
+        if not isinstance(data, xr.Dataset):
+            LOGGER.info("Loading Hazard from file: %s", data)
+            data = xr.open_dataset(data)
+        else:
+            LOGGER.info("Loading Hazard from xarray Dataset")
+
+        # Initialize Hazard object
+        hazard = cls(haz_type=hazard_type)
+        hazard.unit = intensity_unit
+
+        # Update coordinate identifiers
+        coords = copy.deepcopy(DEF_COORDS)
+        coordinate_vars = coordinate_vars if coordinate_vars is not None else {}
+        unknown_coords = [co for co in coordinate_vars if co not in coords]
+        if unknown_coords:
+            raise ValueError(
+                f"Unknown coordinates passed: '{unknown_coords}'. Supported "
+                f"coordinates are {list(coords.keys())}."
+            )
+        coords.update(coordinate_vars)
+
+        # Retrieve dimensions of coordinates
+        dims = dict(
+            event=data[coords["event"]].dims,
+            longitude=data[coords["longitude"]].dims,
+            latitude=data[coords["latitude"]].dims,
+        )
+
+        # Stack (vectorize) the entire dataset into 2D (time, lat/lon)
+        # NOTE: We want the set union of the dimensions, but Python 'set' does not
+        #       preserve order. However, we want longitude to run faster than latitude.
+        #       So we use 'dict' without values, as 'dict' preserves insertion order
+        #       (dict keys behave like a set).
+        data = data.stack(
+            event=dims["event"],
+            lat_lon=dict.fromkeys(dims["latitude"] + dims["longitude"]),
+        )
+
+        # Transform coordinates into centroids
+        hazard.centroids = Centroids.from_lat_lon(
+            data[coords["latitude"]].values, data[coords["longitude"]].values, crs=crs,
+        )
+
+        def to_csr_matrix(array: np.ndarray) -> sparse.csr_matrix:
+            """Store a numpy array as sparse matrix, optimizing storage space
+
+            The CSR matrix stores NaNs explicitly, so we set them to zero.
+            """
+            return sparse.csr_matrix(np.where(np.isnan(array), 0, array))
+
+        # Read the intensity data and flatten it in spatial dimensions
```

Comment:
Data is already flattened
```suggestion
        # Read the intensity data
```